src/coordinator.rs

```diff
@@ -168,41 +172,108 @@ impl Builder {
         ///////////////////////////////////////////////////////////////////////////
-        for glue_job in self.targets.drain(..) {
-            // Note: this data structure is going to grow the ability to
-            // refer to other jobs as soon as it's possibly feasible. When
-            // that happens, a depth-first search through the tree rooted at
-            // `glue_job` will probably suffice.
-            let job =
-                Job::from_glue(glue_job).context("could not convert glue job to actual job")?;
-
-            // TODO: pushing to `ready` immediately is only reasonable when
-            // we don't have job inputs as dependencies, but we don't have that
-            // yet. This'll need to change when we do or we'll have some very
-            // broken runs!
-            coordinator.ready.push(job.base_key);
+
+        // to build a graph, we need the base keys for all jobs. This can't be
+        // a depth-first search, however, because that would mean processing
+        // dependent jobs before their dependencies. We can't do that because
+        // then we would have unresolved base keys and we'd have to defer until
+        // we got them or temporarily suffer incomplete information in the
+        // graph.
+        //
+        // Ideally, we'd look at the leaf nodes first, then the things that
+        // depend on them, etc. In other words, a depth-first search starting
+        // at the leaves instead of the roots. Lucky for us, that's easy to do:
+        // just write down the jobs we see as we do a depth first search, then
+        // walk that list in the opposite direction.
+        //
+        // `to_descend_into` tracks the depth-first search part of this scheme,
+        // and `to_convert` is where we write down the dependencies in root-to-
+        // leaf order.
+        let mut to_descend_into = self.roots.clone();
+        let mut to_convert = Vec::with_capacity(self.roots.len());
+
+        let mut glue_to_job_key: HashMap<&glue::Job, job::Key<job::Base>, Xxh3Builder> =
+            HashMap::with_capacity_and_hasher(self.roots.len(), Xxh3Builder::new());
+
+        let mut job_deps: HashMap<&glue::Job, HashSet<&glue::Job, Xxh3Builder>, Xxh3Builder> =
+            HashMap::with_hasher(Xxh3Builder::new());
+
+        while let Some(next_glue_job) = to_descend_into.pop() {
+            next_glue_job
+                .as_Job()
+                .inputs
+                .iter()
+                .filter(|item| item.discriminant() == glue::discriminant_U1::FromJob)
+                .for_each(|item| {
+                    let job = unsafe { item.as_FromJob() }.0;
+
+                    let entry = job_deps.entry(next_glue_job);
+                    entry
+                        .or_insert_with(|| HashSet::with_capacity_and_hasher(1, Xxh3Builder::new()))
+                        .insert(job);
+
+                    to_descend_into.push(job);
```

Comment:
Can two different jobs depend on the same job? If so, won't this potentially lead to descending into the same job multiple times?

Reply:
yes, they can. Good catch! Gonna have some work to do here, looks like.

Reply:
turns out that descending into the same job is fine as long as we don't get a cycle (and Roc won't let us make a cycle, AFAICT, so it's OK!)

Reply:
(oh, but I stopped it from processing the same job multiple times)